### DISCUSSION TITLE
Changes for inventory_grid.gd regarding create_and_add_item for implementing get_item_at_position

### DISCUSSION THREAD

**Discussion by filipinyo** [peter-kish/gloot#38] [Q&A] [answered]
Hey there. Since create_and_add_item function has been added yesterday, I believe that it needs a separate implementation for the inventory_grid.gd, because when we add and create an item in inventory_grid, it also gets added a position, but in the base inventory, item position is not added.  I've noticed this because I wanted to implement a function that gets an item at the certain position in the inventory. This function would be useful for mapping items to hotkeys etc.

For getting item at the position I've made this function
```
func get_item_at_position(selected_position: Vector2) -> InventoryItem:
	if(
		selected_position.x < 0 || 
		selected_position.x > size.x ||
		selected_position.y < 0 ||
		selected_position.y > size.y
	):
		assert("Selected position out of bounds")

	for item in get_items():
		var item_position = item.get_property(KEY_GRID_POSITION, Vector2.ZERO)
		if (
			item_position.x == selected_position.x && 
			item_position.y == selected_position.y
		):
			return item
	
	return null
```

For overriding the base classes function we can use the same function which calls the inventory_grid's add_item functions that handles position properties.
```
func create_and_add_item(prototype_id: String) -> InventoryItem:
	var item: InventoryItem = InventoryItem.new()
	item.prototype_id = prototype_id
	if add_item(item):
		return item
	else:
		item.free()
		return null
```

If this is already possible in the current implementation, I'd like to know how, so I can use it the way it was meant to be used :D

PS: I really like the plugin

**Comment by peter-kish** [comment, marked as answer]
Yeah, helper functions like `create_and_add_item_at` and `get_item_at` are still missing, however I think it shouldn't be hard to implement them.
One thing you're missing in your implementation is that items with missing `KEY_GRID_POSITION` property are considered to be positioned at `(0, 0)`. This makes the whole thing a lot simpler, as items can be added at certain positions simply by adding them at `(0, 0)` and moving them to the desired grid coordinates:
```gd
func create_and_add_item_at(prototype_id: String, position: Vector2) -> InventoryItem:
    var item = create_and_add_item(prototype_id)
    if item:
        move_item_to(item, position)
    return item
```

Getting items at specific grid coordinates can currently be done with help of `get_item_rect` that takes the missing `KEY_GRID_POSITION` into account:
```gd
func get_item_at(position: Vector2) -> InventoryItem:
    for item in get_items():
        if get_item_rect(item).has_point(position):
            return item
    return null
```

Note: I haven't actually tested any of the above, so I'm not 100% sure it would work.
Note2: Comparing vectors with `==` might not be the best idea since we're effectively comparing `float`s. Yes, we round them to the nearest `int`s but still...

EDIT: Implemented in https://github.com/peter-kish/gloot/pull/40

**Reply by peter-kish** [reply]
I moved the issue to discussions/Q&A to encourage posting similar questions there.
Also created an issue for implementing `create_and_add_item_at` and `get_item_at`: https://github.com/peter-kish/gloot/issues/39
